### PR TITLE
ddl: delay before changing column from null to not null

### LINKED ISSUE #23363: Risk of breaking consistency when changing column from null to not null
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

Initial table:

```sql
CREATE TABLE t (id INT PRIMARY KEY, v INT);
INSERT INTO t VALUES (0, 0);
```
Change the value to `NULL` and change the value to not null concurrently.

```sql
-- Session 1
BEGIN;

-- Session 2 (Concurrently)
ALTER TABLE t MODIFY COLUMN v INT NOT NULL;

-- Session 1 (Concurrently)
UPDATE t SET v = NULL WHERE ID = 0;
COMMIT;
```

### 2. What did you expect to see? (Required)

Either the DDL or the transaction fails.

### 3. What did you see instead (Required)

I think it is possible that both the DDL and the transaction succeed. Because under async commit, there is no check after the not null flag is added.

Then, the table contains NULL, but the column schema is NOT NULL.

I'm still trying to create a test case that can reproduce it.

### 4. What is your TiDB version? (Required)

5.0/nightly

### PR DESCRIPTION
**Pull request by sticnarf**
### What problem does this PR solve?

Issue Number: close #23363

### What is changed and how it works?

Before the double check of changing null to not null, we should delay like the reorg to make async commit safe.

After the delay, if an async commit writes a NULL, the double null check must detect it.

### Check List 

Tests 

- Integration test

### Release note 

- No release note.

### CODE REVIEW

File: ddl/ddl.go [hunk 1]
```diff
@@ -668,6 +668,17 @@ type RecoverInfo struct {
 	CurAutoRandID int64
 }
 
+// delayForAsyncCommit sleeps `SafeWindow + AllowedClockDrift` before a DDL job finishes.
+// It should be called before any DDL that could break data consistency.
+// This provides a safe window for async commit and 1PC to commit with an old schema.
+func delayForAsyncCommit() {
```
Comment (wjhuang2016) [inline ddl/ddl.go:674]: Do we need to check 1PC here?

Comment (sticnarf) [reply ddl/ddl.go:674]: 1PC works in the same way with async commit.

Comment (wjhuang2016) [reply ddl/ddl.go:674]: "sleep before DDL finishes to make async commit safe" is not accurate?

Comment (sticnarf) [reply ddl/ddl.go:674]: OK. I just change the log without changing the function name. It looks too verbose to also put 1PC to the name

**Comment by cfzjywxk** [comment]
@AilinKid @wjhuang2016 PTAL

File: go.sum [hunk 2]
```diff
@@ -5,27 +5,22 @@ cloud.google.com/go v0.44.1/go.mod h1:iSa0KzasP4Uvy3f1mN/7PiObzGgflwredwwASm/v6A
...
 cloud.google.com/go v0.45.1/go.mod h1:RpBamKRgapWJb87xiFSdk4g1CME7QZg3uwTez+TSTjc=
 cloud.google.com/go v0.46.3/go.mod h1:a6bKKbmY7er1mI7TEI4lsAkts/mkhTSZK8w33B4RAg0=
-cloud.google.com/go v0.50.0 h1:0E3eE8MX426vUOs7aHfI7aN1BrIzzzf4ccKCSfSjGmc=
 cloud.google.com/go v0.50.0/go.mod h1:r9sluTvynVuxRIOHXQEHMFffphuXHOMZMycpNR5e6To=
 cloud.google.com/go v0.52.0/go.mod h1:pXajvRH/6o3+F9jDHZWQ5PbGhn+o8w9qiu/CffaVdO4=
 cloud.google.com/go v0.53.0 h1:MZQCQQaRwOrAcuKjiHWHrgKykt4fZyuwF2dtiG3fGW8=
 cloud.google.com/go v0.53.0/go.mod h1:fp/UouUEsRkN6ryDKNW/Upv/JBKnv6WDthjR6+vze6M=
 cloud.google.com/go/bigquery v1.0.1/go.mod h1:i/xbL2UlR5RvWAURpBYZTtm/cXjCha9lbfbpx4poX+o=
-cloud.google.com/go/bigquery v1.3.0 h1:sAbMqjY1PEQKZBWfbu6Y6bsupJ9c4QdHnzg/VvYTLcE=
 cloud.google.com/go/bigquery v1.3.0/go.mod h1:PjpwJnslEMmckchkHFfq+HTD2DmtT67aNFKH1/VBDHE=
```
Comment (AilinKid) [inline go.sum:13]: go.sum should be changed?

**Review by AilinKid — COMMENTED** [review]
Rest LGTM

**Comment by sticnarf** [comment]
https://internal.pingcap.net/idc-jenkins/blue/rest/organizations/jenkins/pipelines/tidb_ghpr_unit_test/runs/68096/nodes/82/steps/345/log/?start=0

Unit test fails due to a reported race: https://github.com/pingcap/tidb/issues/23280

**Pull request merged**